### ISSUE TITLE
Phase 2 결과: CGBD (Consistency-Guided Boundary-Targeted Distillation) 완료

### ISSUE THREAD

**Issue by kimshhhhh** [kimshhhhh/consistency-reasoning#7]
## Phase 2 요약: CGBD Pipeline 전체 실험 결과

**실험 목표**: 학생 모델(Qwen2.5-7B-Instruct)의 **boundary zone 문제만 선별**하여 교사 모델(Qwen2.5-Math-7B-Instruct)로 distillation하면, 랜덤 선별 대비 더 효율적으로 성능이 오르는가?

---

## 실험 설정

| 항목 | 값 |
|------|-----|
| 학생 모델 | Qwen/Qwen2.5-7B-Instruct |
| 교사 모델 | Qwen/Qwen2.5-Math-7B-Instruct |
| 데이터셋 | MATH (Level 4–5, 500문제 샘플) |
| 평가 지표 | pass@1, pass@8, pass@32, boundary graduation rate |
| Zone 기준 | inner: pass@32 > 0.8, outer: < 0.1, boundary: [0.1, 0.8] |
| SFT 방식 | QLoRA (4-bit NF4, LoRA r=16, alpha=32) |
| SFT Epochs | 3, LR=2e-4, effective batch=16 |
| Teacher solutions | 문제당 8개 샘플, correct solution만 사용 |

---

### Step 1: Teacher Solution 생성

교사 모델로 각 조건별 학습 데이터 생성:

| 조건 | 설명 | 문제 수 |
|------|------|---------|
| **cgbd** | Phase 1에서 boundary zone 판정된 문제만 | 110개 |
| **random** | 랜덤 샘플링 (boundary와 동일한 N=110) | 110개 |
| **all** | 전체 500문제 (상한선 baseline) | 500개 |

각 문제에서 적어도 1개 이상의 정답 솔루션을 가진 경우만 SFT에 사용.

### Step 2: QLoRA SFT Fine-tuning

각 조건별로 동일한 하이퍼파라미터로 SFT 진행.  
GPU 메모리 제약(24GB)으로 QLoRA 적용. 학습 완료 후 LoRA adapter 저장.

### Step 3: 평가 (500문제 × 32샘플)

LoRA adapter를 base model에 merge 후 vLLM으로 추론.  
Phase 1과 동일한 500문제에 대해 pass@k 및 zone 재분류.

### Step 4: 분석

Phase 1 baseline 대비 각 조건의 성능 변화, zone 분포 변화, boundary graduation rate 비교.

---

### Pass@k 비교

| 조건 | pass@1 | pass@8 | pass@32 | vs Baseline |
|------|--------|--------|---------|-------------|
| Baseline (no training) | 0.5911 | 0.7455 | 0.8220 | — |
| **CGBD (proposed)** | **0.6016** | **0.7544** | 0.8120 | **+0.0106** |
| Random (baseline) | 0.5781 | 0.7384 | 0.8180 | -0.0130 |
| All (upper bound) | 0.6390 | 0.7720 | 0.8240 | +0.0479 |

### Zone 분포 변화

| 조건 | inner | boundary | outer |
|------|-------|----------|-------|
| Baseline | 50.2% | 22.0% | 27.8% |
| CGBD | 51.2% | 22.8% | **26.0%** |
| Random | 47.6% | 23.2% | 29.2% |
| All | **57.0%** | 18.4% | 24.6% |

### Boundary Zone Graduation Rate

boundary → inner로 이동한 문제 비율 (Phase 1 boundary 110문제 기준):

| 조건 | Graduation Rate |
|------|----------------|
| **CGBD** | **29.1%** |
| Random | 10.9% |
| All | 33.6% |

---

### ✅ 긍정적 신호

1. **CGBD vs Random: Δpass@1 = +0.0235**  
   동일한 110개 문제를 사용했지만, boundary 선별이 랜덤 대비 명확히 우월. 선별 전략 자체의 효과가 있음.

2. **Boundary Graduation Rate: CGBD 29.1% vs Random 10.9%**  
   이게 핵심 메트릭. CGBD는 boundary 문제를 inner로 이동시키는 데 **2.7배** 더 효과적. 이는 CGBD가 targeted training을 의도대로 수행하고 있음을 보여줌.

3. **Random은 오히려 baseline보다 pass@1이 낮음 (-0.013)**  
   랜덤 선별 SFT는 오히려 해가 될 수 있음을 시사. 선별 없는 소량 distillation의 위험성.

4. **All 조건이 가장 높지만** 데이터를 4.5배 더 씀 (500 vs 110). 효율성 면에서 CGBD가 경쟁력 있음.

### ⚠️ 한계 및 주의점

1. **pass@1 절대 향상폭이 작음 (+0.0106)**  
   3 epoch SFT로는 포화 상태에 가까운 base model을 크게 개선하기 어려움. 더 많은 epoch, 더 큰 teacher-student gap이 필요할 수 있음.

2. **pass@32가 CGBD에서 오히려 소폭 감소 (0.822 → 0.812)**  
   SFT 이후 모델이 boundary 문제에 집중하면서 이미 쉬운 문제(inner)에서 약간의 forgetting 가능성.

3. **통계적 유의성 미검증**  
   500문제 기준이지만, 차이가 작아 bootstrap 또는 paired t-test 필요.

4. **Level 4–5 한정 실험**  
   전체 MATH dataset으로 일반화 여부 불명확.

---

## Phase 3 제안: 무엇을 해야 하는가?

Phase 2가 **개념 검증(proof-of-concept)** 수준이라면, Phase 3는 이를 **강화하고 일반화**하는 방향이어야 함.

### Option A: CGBD + Iterative Refinement (권장)

**아이디어**: 한 번 SFT 후 boundary가 이동한 문제를 재분류 → 새 boundary에 다시 SFT → 반복

- Round 1: Phase 2와 동일 (110개 boundary → SFT)
- Round 2: CGBD 모델로 다시 consistency 측정 → 새 boundary 식별 → 교사 솔루션 재생성 → 추가 SFT
- 예상 효과: boundary graduation rate 누적 향상, pass@1 지속 증가

**구현**: Phase 2 pipeline을 루프로 감싸면 됨 (2–3 round 실험)

### Option B: CGBD + DPO (Direct Preference Optimization)

**아이디어**: SFT 이후 boundary 문제에서 correct/incorrect 쌍을 DPO로 추가 정렬

- Teacher correct solution → chosen
- Student wrong solution → rejected
- Boundary 문제에만 집중한 DPO → 더 targeted

**구현**: `trl.DPOTrainer` 사용, Phase 2 teacher data 재활용 가능

### Option C: Consistency Score를 Curriculum Weight로 활용

**아이디어**: SFT 시 consistency score를 sample weight로 사용

- consistency ≈ 0.5 (완전한 boundary) → 높은 weight
- consistency ≈ 1.0 또는 ≈ 0.0 → 낮은 weight
- Loss = Σ w_i × CE_i

**구현**: Custom Trainer, per-sample weight 적용

### Option D: Teacher-Student Gap 확대

**아이디어**: 더 강한 교사 모델 사용 (Qwen2.5-Math-72B 또는 GPT-4o)

- 현재 7B vs 7B → gap이 작아 distillation 신호가 약할 수 있음
- 72B 교사 사용 시 솔루션 품질 대폭 향상 예상

---

## 권장 Phase 3 방향

**Iterative CGBD (Option A)** + **DPO refinement (Option B)** 조합:

1. Round 1–3 iterative SFT (각 round마다 boundary 재측정)
2. 최종 round 후 DPO로 fine-grained alignment
3. 전체 MATH Level 1–5로 평가 범위 확대
4. Statistical significance test 포함

이 방향이 논문 contribution을 가장 명확하게 보여줄 수 있음:  
_"Iterative boundary-targeting + DPO alignment consistently outperforms random distillation and single-round SFT with the same training budget."_

---

## 생성된 파일

```
results/
├── phase2_teacher_cgbd.jsonl      # CGBD 조건 교사 솔루션 (110문제)
├── phase2_teacher_random.jsonl    # Random 조건 교사 솔루션 (110문제)
├── phase2_teacher_all.jsonl       # All 조건 교사 솔루션 (500문제)
├── phase2_eval_cgbd.jsonl         # CGBD 조건 평가 결과 (500문제)
├── phase2_eval_random.jsonl       # Random 조건 평가 결과 (500문제)
├── phase2_eval_all.jsonl          # All 조건 평가 결과 (500문제)
├── phase2_analysis.json           # 최종 수치 요약
├── figures/
│   ├── phase2_zone_distribution.png
│   ├── phase2_pass_at_k.png
│   └── phase2_boundary_graduation.png
└── models/
    ├── cgbd/final/                # CGBD LoRA adapter
    ├── random/final/              # Random LoRA adapter
    └── all/final/                 # All LoRA adapter
```
